stir/src/prover/mod.rs

```diff
@@ -0,0 +1,575 @@
+        log_next_folding_factor,
+        pow_bits,
+        num_queries,
+        num_ood_samples,
+        ..
+    } = config.round_config(round).clone();
+
+    let StirRoundWitness {
+        domain_l,
+        domain_k,
+        evals_k,
+        merkle_tree,
+        folding_randomness,
+        ..
+    } = witness;
+
+    // ================================ Folding ================================
+
+    // Obtain g_i as the folding of f_{i - 1}
+
+    // NP TODO
+    assert_eq!(domain_k.size(), evals_k.len(), "round: {}", round);
+
+    let folded_evals_k =
+        fold_evaluations_at_domain(evals_k, domain_k, log_folding_factor, folding_randomness);
+    let domain_k_pow = domain_k.exp_power_of_2(log_folding_factor).unwrap();
+    let folded_polynomial = domain_idft(folded_evals_k.clone(), domain_k_pow, &config.dft);
+
+    // TODO: Remove this assert and just pad folded_polynomial by zeroes to make it true.
+    // For now this assert seems to hold true for the tests I am running.
+    assert!(folded_polynomial.len() % F::Packing::WIDTH == 0);
+    let packed_folded_polynomial = folded_polynomial
+        .chunks_exact(F::Packing::WIDTH)
+        .map(|chunck| EF::ExtensionPacking::from_ext_slice(chunck))
+        .collect::<Vec<_>>(); // TODO: Ideally this transformation should be a method in ExtensionPacking.
+
+    // Compute the i-th domain L_i = w * <w^{2^i}> = w * (w^{-1} * L_{i - 1})^2
+    // and its subdomain K_{i - 1}
+    let new_domain_l = domain_l.shrink_coset(1).unwrap(); // Can never panic due to parameter set-up
+    let new_domain_k = domain_k.shrink_coset(log_folding_factor).unwrap(); // Idem
+
+    // Evaluate g_i over L_i
+    let folded_evals_l = domain_dft(new_domain_l, folded_polynomial.clone(), &config.dft);
+
+    // Collecting the evaluations of g_i over K_i for later use
+    let g_i_evals_k = folded_evals_l
+        .iter()
+        .step_by(1 << (new_domain_l.log_size() - new_domain_k.log_size()))
+        .copied()
+        .collect_vec();
+
+    // Stack the evaluations, commit to them (in preparation for
+    // next-round-folding verification, and therefore with width equal to the
+    // folding factor of the next round) and then observe the commitment
+    let new_stacked_evals = RowMajorMatrix::new(
+        folded_evals_l,
+        1 << (new_domain_l.log_size() - log_next_folding_factor),
+    )
+    .transpose();
+
+    let (new_commitment, new_merkle_tree) = config.mmcs_config().commit_matrix(new_stacked_evals);
+
+    // Observe the commitment
+    challenger.observe(F::from_u8(Messages::RoundCommitment as u8));
+    challenger.observe(new_commitment.clone());
+
+    // ======================== Out-of-domain sampling ========================
+
+    let mut ood_samples = Vec::new();
+
+    challenger.observe(F::from_u8(Messages::OodSamples as u8));
+    while ood_samples.len() < num_ood_samples {
+        let el: EF = challenger.sample_algebra_element();
+        if !new_domain_l.contains(el) {
```

Comment:
I get what the goal of this is but this check is non trivially expensive (It involves a bunch of extension field multiplications).

How about replacing it with just checking if it's in the base field:
```suggestion
        if !el.is_in_basefield() {
```
This has a slightly higher failure rate (|F|/|EF|) but it should basically always pass.

Reply:
The whole protocol is happening over EF though, and in particular that's where the domains L_i live (on the verifier, cf. the call to `EF::two_adic_enerator`; in `commit_polynomial` and `commit_evals`, the `F` in `F::two_adic_generator` is `EF` in practice, but the function has a single field-related generic, which is simply called `F`). So  `!el.is_in_basefield()` doesn't guarantee what we need, `!new_domain_l.contains(el)`. Am I missing something?

Consider Goldilocks, which has a 2-adicity of 32, but whose every quadratic extension has a two-adicity of 33. One could run the STIR for, say, degree 2^30 and inverse rate 2^3 over the quadratic extension in Plonky3, and the domain L_0 wouldn't be contained in Goldilocks. 

As a second, maybe less clear-cut argument against the change, I think it would be nice if the protocol can run over large prime fields (e. g. Fp for p ~ 2^256) instead of always over extensions of small prime fields (as it is now, e. g. a quintic extension of BabyBear). In that case, `f !el.is_in_basefield()` would always fail. For instance, the benchmarks in the original STIR paper are run over such large prime fields. In any case you, as a maintainer of the repo, are better equipped to judge whether such fields are interesting to cover at all.

Reply:
Yeah that's a fair point. I was assuming that the set up would be such that the domain is always in the base field. (I think that might be a reasonable constraint to impose for performance reasons).

We actually found a similar issue about ood samples in Plonky3 (see #818). After discussions, we think the best bet is to just remove this check from the prover, but leave it in for the verifier and accept that this leads to a small completeness error.

(There is actually already a non trivial completeness error coming from how PoW works so having a little more completeness error shouldn't be an issue)

Reply:
Is that check really so costly? Note that `ood_samples` is typically 1 or 2. If there is a round with codeword length `|L| = 2^N`, then for `ood_samples = 2`, all `new_domain_l.contains(el)` does is perform at most `2 * 2 * N` squarings (in the field extension). Compare this with the several FFTs and iFFTs in the same round, at least one of which is over a all of `L` (with a cos oft `~ N * 2^N` multiplications using the naive implementation).

If you think so, we're happy to remove the check and perhaps leave a small comment clarifying the decision in the code. To be clear, which of the two versions of the check would you keep in the verifier? The more general (but costlier) one `if !new_domain_l.contains(el)`?

If we do remove the original check is removed on the prover, it would probably be clean to add an extra check somewhere else, as it could then can happen (if I'm not mistaken) that some denominator vanished. It would likely be clean to detect that and handle it somehow instead of having an uncontrolled `panic`. What would you propose? Having `prove` return `Result<Proof>` seems clean. Otherwise one could e. g. return an empty `Proof`.

Over on the verifier side, the check would be left in place and, if not passed, the verifier would simply reject the proof at that point. How does this sound?

Reply:
I totally agree that this isn't a big cost, but I think there are `2` advantages in removing it.
- 1: Slight time saving.
- 2: Makes it easier to do recursion on STIR. A comment we have gotten from Succinct is that loops which require an unknown number of random samples are painful for recursion. That was the main reason we didn't add this into the Plonky3 prover.

(I think leaving a comment is a good idea if we do remove it)

To be completely honest, I think it would be totally okay to just not check it, assume it does not lie in the domain and accept a slightly higher chance of an uncontrollable panic (Particularly given how low this chance is). Whenever you get a proof of work witness using `.grind` there is a `~ 2^{\# bits}/(|F|^2)` of a prover panic where `|F|` is the base field. E.g. if you are working over an extension field of `KoalaBear` and want a 20bit pow, there is a `1/2^{42}` probability of a panic.

I think this probability is a higher than the risk of the sample falling into the domain.

On the verifier side, I agree they should check this and reject if it doesn't hold or, possibly, just reject if they ever are asked to divide by 0. Would that be enough?